Cargill/splinter
# Add profile provider and biome/profile REST API endpoints

Create a ProfileProvider trait and implement it for github and openid. The trait has a method 'get_profile' which takes an access token as an argument and queries the user info endpoint to retrieve the profile information given by the oauth provider.

Add ProfileProvider to the OAuthClient struct. Add a 'with_profile_provider' method to the OAuthClientBuilder as well as the github and openid OAuthClientBuilders.

Add a 'profile' field to the UserInfo struct. Modify the 'exchange_authorization_code' method to retrieve the profile information
using the profile provider and set the profile field of the UserInfo struct when the biome-profile feature is enabled.

Modify the 'make_callback_route' method to take a UserProfileStore as an argument and save the profile information stored in the profile field of the UserInfo struct if the biome-profile feature is enabled. If the given profile already exists in the profile store it will be updated otherwise it will be added as a new profile.

Add a user_profile_store field to AuthConfig::OAuth and OAuthResourceProvider these fields are only available when the 'biome-profile' feature is enabled.

Impl From oauth::Profile for biome::profile::store::Profile to allow for conversion between the two structs.

Add the biome/profile/{id} and biome/profiles endpoints for retrieving and listing biome user profiles.


## **Testing:**
1. Start splinterd with an oauth provider
```
$ docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=mypassword -e POSTGRES_DB=splinter -d postgres
$ cargo run --manifest-path cli/Cargo.toml -- database migrate -C postgres://postgres:mypassword@localhost:5432/splinter
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- cert generate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features experimental -- \
  --database postgres://postgres:mypassword@localhost:5432/splinter \
  --oauth-provider <provider> \
  --oauth-client-id <client-id> \
  --oauth-client-secret <client-secret> \
  --oauth-redirect-url http://localhost:8080/oauth/callback \
  --rest-api-endpoint http://localhost:8080
  --tls-insecure
  -vv
```

2. In a browser go to `localhost:8080/oauth/login?redirect_url=redirect`

3. Go back to the terminal window where splinterd is running and check that it shows `INFO [splinter::oauth::rest_api::actix::callback] User profile saved`
4. Copy the access token from splinterd, it should be in a log message: `GET /oauth/redirect?access_token=<access-token>` and Make the following curl calls

`curl -i -H'Authorization:Bearer <access-token>' http://localhost:8080/biome/profile`
should return the profile information for the profile you used to log in

copy the user_id from the _Authenticated user_ debug message from splinterd 
`curl -i -H'Authorization:Bearer <access-token>' http://localhost:8080/biome/profiles/<user_id>`
should get a `HTTP/1.1 401 Unauthorized` indicating the endpoint exists

`curl -i -H'Authorization:Bearer <access-token>' http://localhost:8080/biome/profiles`
should get a `HTTP/1.1 401 Unauthorized` indicating the endpoint exists
